src/coreclr/jit/optimizer.cpp

```diff
@@ -4564,6 +4564,346 @@ PhaseStatus Compiler::optUnrollLoops()
+        if (middleNext == nullptr)
+        {
+            return false;
+        }
+
+        if (middleNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            foundMiddle = true;
+        }
+
+        // Check that we have linear flow and are still in the same EH region
+
+        if (middleBlock->GetUniquePred(this) == nullptr)
+        {
+            return false;
+        }
+
+        if (!BasicBlock::sameEHRegion(middleBlock, block))
+        {
+            return false;
+        }
+
+        // Can all the nodes within the middle block be made to conditionally execute?
+        for (Statement* const stmt : middleBlock->Statements())
+        {
+            GenTree* tree = stmt->GetRootNode();
+            switch (tree->gtOper)
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->gtGetOp1();
+                    GenTree* op2 = tree->gtGetOp2();
+
+                    // Only one per assignment per block can be conditionally executed.
+                    if (asgNode != nullptr || op2->OperIs(GT_SELECT))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the destination of the assign is a local variable with integer type.
+                    if (!op1->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op1))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the nodes of the assign won't cause any additional side effects.
+                    if ((op1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+                        (op2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+
+                    asgNode = tree;
+                    asgStmt = stmt;
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    if (tree->gtGetOp1() != nullptr || (tree->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+    }
+    if (asgNode == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+    // Calculate size costs.
+    // First reclaculate existing costs to make sure they are up to date.
+    gtSetEvalOrder(last);
+    gtSetEvalOrder(asgNode);
+    // If the condition is against 0 then JTRUE will have no size cost as it'll be merged into the compare.
+    int currentCostSz =
+        asgNode->GetCostSz() + (cond->gtGetOp2()->IsIntegralConst(0) ? cond->GetCostSz() : last->GetCostSz());
```

Comment:
Ideally we would fix the costing instead. We should not be working around incorrect costing like this.

Reply:
It's not clear to me that this is really the heuristic we want, barely anything in the JIT uses code size estimates when optimizing for speed (the places that do are more about code duplication -- which this transformation does not do).

Reply:
I would stick with Andy's suggestion here -- put a constant limit on `GetCostEx()` for the code that is now being evaluated eagerly. We can adjust the heuristic as needed once we get some mileage on the transformation.

Reply:
Agreed optimising for speed is better than size.

But - by optimising for speed I always see regressions via spmidiff. That's due the costings not taking into account that comparison against 0 will get optimised away into cbnz. 

With everything enabled:
[14:50:56] Total bytes of base: 138017016 (overridden on cmd)
[14:50:56] Total bytes of diff: 138022692 (overridden on cmd)
[14:50:56] Total bytes of delta: 5676 (0.00 % of base)
[14:50:56] 567 total methods with Code Size differences (87 improved, 480 regressed), 783 unchanged.

Limiting via costEx:
[14:54:47] Total bytes of base: 138017992 (overridden on cmd)
[14:54:47] Total bytes of diff: 138021404 (overridden on cmd)
[14:54:47] Total bytes of delta: 3412 (0.00 % of base)
[14:54:47] 281 total methods with Code Size differences (42 improved, 239 regressed), 572 unchanged.


This is using the following for costing:
```
    // Using SELECT nodes means that full assignment is always evaulated.
    // Put a limit on the original source and destination of the assignment.
    int costing = asgNode->gtGetOp1()->GetCostEx() + asgNode->gtGetOp2()->GetCostEx() - 2;
    JITDUMP("costing=%d\n", costing);

    if (costing > 0 && !compStressCompile(STRESS_IF_CONVERSION, 0))
    {
        JITDUMP("Aborting If Conversion: optimisation too expensive (+%d)\n", costing);
        return false;
    }
```

But, if speed is what we're looking for, then yes, checking spmidiff isn't that useful

Reply:
> That's due the costings not taking into account that comparison against 0 will get optimised away into cbnz.

Sounds like something we ought to fix, if you have free cycles can you submit a separate PR for it?

> `    int costing = asgNode->gtGetOp1()->GetCostEx() + asgNode->gtGetOp2()->GetCostEx() - 2;`

Evaluating the cost of the LHS of assignments uncontextually is odd since it is treated specially ("no cost" if it ends up being a local in a register, for example). `gtSetEvalOrder` uses `gtIsLikelyRegVar` to estimate this, does it make sense to take it into account? Basically, in that case the cost of `ASG(lcl, SELECT(...))` is only going to be the cost of `SELECT` when evaluated into a target register.

I was really just thinking of something like:
```cpp
if (!compStressCompile(STRESS_IF_CONVERSION, weight))
{
  int writeCost = gtIsLikelyRegVar(asgNode->gtGetOp1()) ? 0 : 2;

  if (writeCost + asgNode->gtGetOp2()->GetCostEx() > 3)
  {
    JITDUMP("Skipping if-conversion that will evaluate expensive RHS unconditionally");
    return false;
  }
}
```

Taking `cbnz` into account as you mentioned would be great. Exact values need to be adjusted, for example maybe pick the number such that we allow conversion for `x = cond ? a + b : x`, but no more than that. By the way, clang seems really aggressive: https://godbolt.org/z/54GMGaTfY

I'm not that worried about the size wise regressions.

Reply:
>Sounds like something we ought to fix, if you have free cycles can you submit a separate PR for it?

Sure, wasn't sure if it was intentional. I'll do that and the incorrect size one too.

>I was really just thinking of something like:

Will try this out.

> By the way, clang seems really aggressive: https://godbolt.org/z/54GMGaTfY

I'm not certain on this, but, I think in later passes clang will sometimes turn the if conversions back into branches (maybe as part of auto vec).

Reply:
Updated costings and set the Stress to 25%.

Limit of 7 to allow for:
```
N005 (  7,  5) [000016] -A------R--                         *  ASG       int    $VN.Void
N004 (  1,  1) [000015] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
N003 (  7,  5) [000014] -----------                         \--*  ADD       int    $102
N001 (  3,  2) [000012] -----------                            +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
N002 (  3,  2) [000013] -----------                            \--*  LCL_VAR   int    V03 arg3         u:1 (last use) $83
```